storage/virtio-blk/vblksrv.c

```diff
@@ -0,0 +1,557 @@
+	}
+
+	return -EINVAL;
+}
+
+
+static ssize_t vblksrv_write(storage_t *strg, off_t off, const void *buf, size_t size)
+{
+	if ((strg == NULL) || (strg->dev == NULL) || ((off + size) > strg->size)) {
+		return -EINVAL;
+	}
+
+	if (size == 0) {
+		return 0;
+	}
+
+	storage_blk_t *blk = strg->dev->blk;
+	if ((blk != NULL) && (blk->ops != NULL) && (blk->ops->write != NULL)) {
+		return blk->ops->write(strg, off, buf, size);
+	}
+
+	return -EINVAL;
+}
+
+
+static int vblksrv_getAttr(storage_t *strg, int type, long long *attr)
+{
+	if ((strg == NULL) || (attr == NULL)) {
+		return -EINVAL;
+	}
+
+	switch (type) {
+		case atSize:
+			*attr = (off_t)strg->size;
+			break;
+
+		default:
+			return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+/* Message handler */
+
+
+static void vblk_msgHandler(void *arg, msg_t *msg)
+{
+	storage_t *strg;
+	mount_i_msg_t *imnt = (mount_i_msg_t *)msg->i.raw;
+	mount_o_msg_t *omnt = (mount_o_msg_t *)msg->o.raw;
+
+	(void)arg;
+
+	switch (msg->type) {
+		case mtOpen:
+		case mtClose:
+			strg = storage_get(msg->i.openclose.oid.id);
+			msg->o.io.err = (strg == NULL) ? -EINVAL : EOK;
+			break;
+
+		case mtRead:
+			strg = storage_get(msg->i.io.oid.id);
+			msg->o.io.err = vblksrv_read(strg, msg->i.io.offs, msg->o.data, msg->o.size);
+			break;
+
+		case mtWrite:
+			strg = storage_get(msg->i.io.oid.id);
+			msg->o.io.err = vblksrv_write(strg, msg->i.io.offs, msg->i.data, msg->i.size);
+			break;
+
+		case mtSync:
+			msg->o.io.err = -ENOSYS;
+			break;
+
+		case mtGetAttr:
+			strg = storage_get(msg->i.attr.oid.id);
+			msg->o.attr.err = vblksrv_getAttr(strg, msg->i.attr.type, &msg->o.attr.val);
+			break;
+
+		case mtMount:
+			strg = storage_get(imnt->dev.id);
+			omnt->err = storage_mountfs(strg, imnt->fstype, msg->i.data, imnt->mode, &imnt->mnt, &omnt->oid);
+
+		case mtUmount:
+			strg = storage_get(((oid_t *)msg->i.data)->id);
+			omnt->err = storage_umountfs(strg);
+			break;
+
+		case mtMountPoint:
+			strg = storage_get(((oid_t *)msg->i.data)->id);
+			omnt->err = storage_mountpoint(strg, &omnt->oid);
+			break;
+
+		default:
+			msg->o.io.err = -EINVAL;
+			break;
+	}
+}
+
+
+/* Device initialization */
+
+
+static int vblksrv_devInit(storage_t *strg, struct _storage_devCtx_t *ctx)
+{
+	if (strg->parent == NULL) {
+		strg->start = 0;
+		strg->size = ctx->size;
+	}
+
+	/* Initialize dev structure */
+	strg->dev = malloc(sizeof(storage_dev_t));
+	if (strg->dev == NULL) {
+		return -ENOMEM;
+	}
+
+	/* Initialize BLK interface */
+	strg->dev->blk = malloc(sizeof(storage_blk_t));
+	if (strg->dev->blk == NULL) {
+		free(strg->dev);
+		strg->dev = NULL;
+		return -ENOMEM;
+	}
+
+	strg->dev->blk->ops = vblk_getBlkOps();
+
+	/* No MTD interface */
+	strg->dev->mtd = NULL;
+
+	ctx->req = vblk_open();
+	strg->dev->ctx = ctx;
+
+	return EOK;
+}
+
+
+static void vblksrv_devDestroy(storage_t *strg)
+{
+	if (strg->dev != NULL) {
+		if (strg->dev->blk != NULL) {
+			free(strg->dev->blk);
+		}
```

Comment:
It's ok to `free(NULL)`
```suggestion
		free(strg->dev->blk);
```